pydantic/_internal/_model_construction.py

```diff
@@ -535,3 +538,42 @@ def generate_model_signature(
         merged_params[var_kw_name] = var_kw.replace(name=var_kw_name)
 
     return Signature(parameters=list(merged_params.values()), return_annotation=None)
+
+
+class _PydanticWeakValueDictionary(weakref.WeakValueDictionary):
+    pass
+
+
+def build_lenient_weakvaluedict(d: dict[str, Any] | None) -> dict[str, Any] | None:
+    """Takes an input dictionary, and produces a new value that (invertibly) replaces the values with weakrefs.
+
+    We can't just use a WeakValueDictionary because many types (including int, str, etc.) can't be stored as values
+    in a WeakValueDictionary.
+
+    The `unpack_lenient_weakvaluedict` function can be used to reverse this operation.
+    """
+    if d is None:
+        return None
+    result = {}
+    for k, v in d.items():
+        try:
+            proxy = _PydanticWeakValueDictionary(ref=v)
```

Comment:
If there is an issue using WeakValueDictionary could this instead use `weakref.ref` for `proxy` instead of wrapping it to a whole `_PydanticWeakValueDictionary` with a single key?

Reply:
Something like this would also work by implementing MutableMapping (without overriding WeakValueDictionary and keeping `__getitem__` to return the actual value instead of weakrefs from `__pydantic_parent_namespace__`):
```python
class ParentNamespaceDictionary(MutableMapping[str, Any]):
    def __init__(self) -> None:
        self._data: dict[str, Union[weakref.ReferenceType[Any], Any]] = {}
        self._needs_remove_dead = False

    def __setitem__(self, key: str, value: Any) -> None:
        self._remove_dead()
        try:
            self._data[key] = weakref.ReferenceType(value, self._on_item_finalized)
        except TypeError:
            self._data[key] = value  # Eg str instances

    def __getitem__(self, key: str) -> Any:
        self._remove_dead()
        val = self._data[key]
        if not isinstance(val, weakref.ReferenceType):
            return val
        if (unwrapped := val()) is not None:
            return unwrapped
        raise KeyError(key)

    def __delitem__(self, key: str) -> None:
        self._remove_dead()
        del self._data[key]

    def __len__(self) -> int:
        self._remove_dead()
        return len(self._data)

    def __iter__(self) -> Iterator[str]:
        self._remove_dead()
        return iter(self._data)

    def _remove_dead(self) -> None:
        if self._needs_remove_dead:
            self._needs_remove_dead = False
            dead_keys = [k for k, v in self._data.items() if isinstance(v, weakref.ReferenceType) and v() is None]
            for k in dead_keys:
                del self._data[k]

    def _on_item_finalized(self, _ref: weakref.ReferenceType) -> None:
        self._needs_remove_dead = True
```

Reply:
I see how the `ParentNamespaceDict` would work, but ultimately I think it's a bit more complexity than just replacing the WeakValuesDict with a custom `weakref.ReferenceType` subclass.

I'd be open to a PR making this change. And would be open to doing one myself if there was more of a clear benefit for it, probably in a future where we do more with the parent namespace dict. I'd rather not worry about getting this exactly right though unless that dict gets used more broadly than it is now.

If you disagree I could put in the effort to implement it this way.